wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt

```diff
@@ -115,321 +82,64 @@ class NotificationsAdapter(val context: Context, val wallet: Wallet, val onConta
     }
 
     override fun getItemId(position: Int): Long {
-        return when (results[position].viewType) {
-            NOTIFICATION_NEW_HEADER -> 1L
-            NOTIFICATION_NEW_EMPTY -> 2L
-            NOTIFICATION_CONTACT_REQUEST_RECEIVED -> getLongValue(results[position].notificationItem!!.id)
-            NOTIFICATION_EARLIER_HEADER -> 3L
-            NOTIFICATION_CONTACT_ADDED -> getLongValue(results[position].notificationItem!!.id)
-            NOTIFICATION_PAYMENT -> getLongValue(results[position].notificationItem!!.id)
-            else -> throw IllegalArgumentException("Invalid viewType ${results[position].viewType}")
-        }
-    }
-
-    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
-        when (results[position].viewType) {
-            NOTIFICATION_CONTACT_REQUEST_RECEIVED,
-            NOTIFICATION_CONTACT_ADDED -> holder.bind(results[position].notificationItem!!.usernameSearchResult!!, results[position].isNew)
-            NOTIFICATION_NEW_HEADER -> (holder as HeaderViewHolder).bind(R.string.notifications_new)
-            NOTIFICATION_EARLIER_HEADER -> (holder as HeaderViewHolder).bind(R.string.notifications_earlier)
-            NOTIFICATION_NEW_EMPTY -> (holder as ImageViewHolder).bind(R.drawable.ic_notification_new_empty, R.string.notifications_none_new)
-            NOTIFICATION_PAYMENT -> (holder as TransactionViewHolder).bind(results[position].notificationItem!!.tx!!)
-            else -> throw IllegalArgumentException("Invalid viewType ${results[position].viewType}")
+        return when (getItemViewType(position)) {
+            NOTIFICATION_HEADER -> 1L
+            NOTIFICATION_EMPTY -> 2L
+            else -> getLongValue(getItem(position).notificationItem.getId())
```

Comment:
This is the only place that has a conflict with my PR #475, which is pretty impressive.  there are some imports, so the merge process will very easy.

Reply:
New polymorphic data model allowed me to simplify this method even more :)

Also, I used `UUID.randomUUID()` for NOTIFICATION_HEADER `ID` instead of hardcoded number because in `NotificationActivity` we have two items of this type (New/Earlier) whereas `ID` should be unique as we as we use the `setHasStableIds(true)`. Previously, when there was a separate type NOTIFICATION_EARLIER_HEADER the uniqueness was assured by hardcoded values. One of advantage of the new approach is that we can have as much header items as we want without the need of creating new item type for each.